### DISCUSSION TITLE
Poor performance with Edge Functions

### DISCUSSION THREAD

**Discussion by tadhglewis** [supabase/supabase#29301] [Questions] [answered]
Hi!

I have been trialing Edge Functions and found the performance pretty awful. This has ranged from ~1.2s cold starts to ~400-200ms with repeated requests (sending 20+ requests within 1 minute) and overall inconsistent and high latency.

This is in comparison to using Deno Deploy (which Edge Functions uses under the hood) directly which has cold starts of ~120ms and ~30ms repeated consistently - similar for Cloudflare Workers.

The above results has been when using a minimal hello-world with zero dependencies
```JavaScript
import "jsr:@supabase/functions-js/edge-runtime.d.ts"

console.log("Hello from Functions!")

Deno.serve(async (req) => {
  const { name } = await req.json()
  const data = {
    message: `Hello ${name}!`,
  }

  return new Response(
    JSON.stringify(data),
    { headers: { "Content-Type": "application/json" } },
  )
})
```
The deploy log is as follows
```
Bundling Function: latency-test
Download https://jsr.io/@supabase/functions-js/meta.json
Download https://jsr.io/@supabase/functions-js/2.4.3_meta.json
Deploying Function: latency-test (script size: 2.796kB)
```

If I use additional dependencies this also leads to _even worse_ performance and increase the latency to ~2s cold start and ~600-400ms repeated requests with minimal additional packages.

Additionally, I have tried different ISPs, locations (different countries in APAC region) as well as manually setting the `x-region` header which all have resulted in the same high and inconsistent latency.

I have reached out to Supabase support and they haven't been very helpful so far so thought I'd chuck it up here if anyone has any ideas or has the same issue as me.

**Comment by tadhglewis** [comment]
Still no help from Supabase support and I've been bounced around three different support engineers with every reply. Not exactly a good impression.

I can see the boot time for this function is ~20ms yet the network latency ranges to as low as ~600ms on cold start.

I am very confident this is **not** a network issue on my end considering I've tried different ISPs, countries, etc. **So why does Suapbase have a internal network latency of over 500ms somewhere in their stack??**
<img width="264" alt="Screenshot 2024-09-18 at 8 06 38 pm" src="https://github.com/user-attachments/assets/62af85a8-0d3e-462e-9663-4d23573af37e">

<img width="216" alt="Screenshot 2024-09-18 at 8 07 04 pm" src="https://github.com/user-attachments/assets/b6e3c298-86e6-44d9-972b-445d7db7a60d">

**Comment by tadhglewis** [comment]
Other people seem to have this issue as well?

> Supabase itself appears to be in excess of 1 full second between ingress and egress by my measurements
https://www.reddit.com/r/Supabase/comments/1fgu7bb/comment/ln6spxa

> Yes, we are currently working on improving the Gateway overhead and have seen some promising results. Will provide an update as we ship.
Reply from @lakshan-supabase

Seems insane to me this hasn't been raised as a production incident and instead I've been bouncing round support staff.

**Comment by tadhglewis** [comment]
To further demonstrate this further, here's two equivalent functions: uses a router, fetches data from Supabase, parse and format it and return JSON

Edge Functions 
<img width="288" alt="Screenshot 2024-09-19 at 12 18 00 am" src="https://github.com/user-attachments/assets/9d55f90e-18ca-4fad-b6a9-4c4fa93cc855">


Cloudflare Workers
<img width="292" alt="Screenshot 2024-09-19 at 12 17 15 am" src="https://github.com/user-attachments/assets/70d19925-142e-4088-ab25-8d16dc4e59fd">

This is repeated requests to ensure neither services were cold starts.

**Comment by tadhglewis** [comment]
It seems like they are self hosting edge functions now instead of using deno deploy according to: https://news.ycombinator.com/item?id=38622830.

I assume this is the root cause of the issue.

**Comment by codinronan** [comment]
Hi @tadhglewis we meet again. I'm going to drop some of my own analysis here, because it's too much to put in a support ticket (which I am also opening, with a truncated version of what I am sharing here, and pointing them here to get further context).

TL;DR of my thesis: We've now run into instances where this gateway overhead is becoming a problem, and I've collected some data to demonstrate it, and there's seemingly little I can do to reduce it. It also seems like whatever optimization Supabase's hosted edge runtime instances are trying to make, it only works at high throughput/continuous traffic. Spikey workloads seem to suffer consistently poor overhead.

The following set of screenshots are a sample of what we are seeing from the browser's point of view and measured using open telemetry. This is one of our production endpoints that does nothing more than a very quick DB lookup and returns. From our function's point of view (using otel) it takes 56ms to execute; from Supabase's point of view it appears to take 440ms to execute (the difference I suppose is boot time?). From the browser's point of view it takes _a full 3 seconds_. We know that roughly 200-300ms of this is our Cloudflare proxy.

[prod endpoint, 3sec execution time, browser dev tools](https://monosnap.com/file/8YEq1Ji9KjxVFFbIjugdla2CLI6Q4S)
[edge function invocation log](https://monosnap.com/file/3M99joN4HFqqOpAVejT56pgZFBVyBh)
[open telemetry trace](https://monosnap.com/file/FNk3yOADfEXXDiWQ5TJkinZvwB3hly)

To test this in a more general way, I did 2 things:
1. Duplicated @tadhglewis very simple no-dependency function above and deployed it to our production instance under the path '/perf'. We will call this **"F1-Unloaded"** 
2. Made a "fully loaded" endpoint for testing. I simplified one of our actual endpoints down to only the imports required for routing, and importing some of our lookup functions (not calling them, just importing them so they are part of the bundle) and logging them to the console (eg "createClient: [Function: Z]," for the supabase create client function). We will call this **"F2-FullyLoaded"**

For testing, I used the Artillery load testing package (cli only). I tested this via a local test and via tests from a Lambda hosted in AWS, region us-east-1. I'm presenting data from both sets of tests, from both of the above endpoints. A truncated artillery config I used looks like:

```yml
config:
  target: https://ourapp.com/api (proxied by cloudflare)
  # target: https://ourprojectid.supabase.co (not proxied, raw to gateway)
  phases:
    - duration: 20
      arrivalRate: 1
      rampTo: 3
      name: Warm up phase
    - duration: 30
      arrivalRate: 2
      rampTo: 8
      name: Ramp up load
  plugins:
    metrics-by-endpoint: {}
scenarios:
  - flow:
      - loop:
          # - get:
          #     url: '/api/functions/v1/perf' (Function 1, no deps)
          - get:
              url: '/functions/v1/utilities/perfsync' (function 2, loaded w/imports)
        count: 50
```

The output of these tests is the following. You can skip the images, I've summarized them below.
Lower load for calibration (ramps only go from 1->2):
1. [F1-Unloaded, run locally, raw endpoint](https://monosnap.com/file/kfkpsw2g20jYsEOmIEPfrzaNpMGW8R)
2. [F1-Unloaded, aws-lambda/us-east-1, raw endpoint](https://monosnap.com/file/9dMPVrFKTURpXn4KE5QqxTSjL36HTt)

Higher load (ramps as you see above in the config):
1. [F1-Unloaded, run locally, raw endpoint](https://monosnap.com/file/5ieqpS7DHd8ArgCzX6gnsg8kUdcAFC)
2. [F2-FullyLoaded, run local, raw endpoint](https://monosnap.com/file/GGFB7QcdiAVVelWSv5o2clgwnz9yZR)
3. [F2-FullyLoaded, aws-lambda/us-east-1, raw endpoint](https://monosnap.com/file/KhoDCAuqG4bn4rnB1t4CzB60nreKj3)
4. [F2-FullyLoaded, run locally, through CF proxy](https://monosnap.com/file/cx2n25BsZ8EwuYSqC0BkGgGMx2nJ2F)
5. [F2-FullyLoaded, aws-lambda/us-east-1, through CF proxy](https://monosnap.com/file/kQuRQg09pQUqXIFAnqaJAFx2f8pZ3G)

As you'd expect the testing from the lambda was able to achieve a higher requests per second. Each test above did between 3500-10000 requests. What's interesting is that the very first link has 502's and ETIMEOUTs which doesn't really make sense, that function is as lightway as is physically possible and the load on that test was very low.

It does seem like the functions actually hold up better under consistent high load, than under spikier, lower load. You can see this in the first image, which will be quite a spikey load, which has worse performance than a fully loaded function under high load called from within AWS (image 3 in the second section, above). This would also be the case for the browser hit I initially shared.

The reason this is becoming an issue for us is that one of our particular use cases requires fanning out the calculation of a piece of customer data across several invokes of the same function. For low-volume clients, this seems ok, but for ones with more data (requiring more fanning) we quickly start to feel the effect of the above (this is essentially a mini map-reduce). Even a cache to avoid recalculation is only giving us a slight boost, because the core issue is that under a bigger workload, even waiting for the cached value to return can take several seconds (this is one edge function calling another edge function in the same project, leaving user networks out of the equation entirely).

**Comment by anton-kirschhock** [comment]
We also notice that our functions (which are not super-complex) are fairly slow when they don't see constant traffic - Hope we get some insights from the Supabase team soon!

**Comment by laktek** [comment, marked as answer]
Hey, want to share we've deployed some changes that should help improve the roundtrip latency.

Based on internal metrics, the cold latency median (first request to a function in an hourly window) is 400ms, and the hot latency median (subsequent requests in the same hourly window) is 125ms.

This could slightly vary across regions, but generally should stay +/- 100ms from the median.

We plan to release more telemetry for Edge Functions in the future, which should give you an accurate breakdown of your functions' latency bottlenecks.

**Reply by tadhglewis** [reply]
Is that released yet? I'm still seeing 2.5s cold start with very spikey hot times ~200ms, ~400ms, ~2s

Having accurate telemetry in the Supabase dashboard would be cool - that way it's more easily measurable

**Reply by laktek** [reply]
Yes, changes have already been released in all regions.

Depending on what your function does at the boot (the code that runs before it's ready to serve the request), cold start can increase. Does it make any network requests or set up any libraries? If you are using NPM modules, try republishing your functions to see if it reduces the cold starts https://supabase.com/blog/edge-functions-faster-smaller 

Also, can you make a GET request to this URL and report the response times you see? https://obuldanrptloktxcffvn.supabase.co/functions/v1/health-check

**Reply by tadhglewis** [reply]
@laktek 

> Does it make any network requests or set up any libraries

No, it's just the hello world example in the original post. I've also tried republishing

> Also, can you make a GET request to this URL and report the response times you see? [obuldanrptloktxcffvn.supabase.co/functions/v1/health-check](https://obuldanrptloktxcffvn.supabase.co/functions/v1/health-check)

I'm also seeing 600ms, 345ms, 200ms, 280ms, 495ms on your health check endpoint similar to my hello world edge function - not sure about cold starts for the health check as other people might be hitting it

So overall I haven't seen much of an improvement and latency is generally pretty far behind Deno Deploy and CF Workers

@codinronan have you noticed any improvements with your tests? 🤔

**Reply by laktek** [reply]
@tadhglewis Can you forward me the previous support ticket to lakshan [at] supabase.com. Knowing your project, regions it accessed from, and other network conditions would help us to figure out why you see increased latency.

**Comment by jpohhhh** [comment]
TL;DR: Things are looking great, I don't see any speed difference between Cloudflare and Supabase, if there is, its < 100 ms AFAICT. If you stumble across this thread and are trying to figure out if you need to take some action, I strongly believe you don't.

Longer version:

My experience has been 99% mobile dev for the last 12 years, and over the last year of building out a startup, I've *heavily* relied on Supabase to avoid having to learn server stuff.

I'm getting close to release, and about 6 weeks ago, started looking into various GitHub and Reddit threads about edge functions being slow. 

I instrumented all my edge functions with Sentry traces, giving me precise breakdowns of slow spans of code.

I don't have a ton of users or huge set of data, its just me, the CI server, and a handful of people, definitely < 10. That being said, it's very clear to me at this point that Cloudflare has marginal speed gains, at best, in the context of proxying API calls to 3rd party services. Think, calling OpenAI, or a search result API. I'm talking about the search results API in the following, because its very steady traditionally, not much variance.

The very latest changes mentioned kept P25 response times ~the same, P50 got probably a 150 ms boost, P75, 200, broadening out to about 600 ms improvement at P99.

I just spent a couple hours wrapping up my first Cloudflare edge function that also does search. My instinct, with some charity to the above posts, is that if I deployed this I might see as much as a 100 ms improvement. That being said, there's no meaningful difference in speed in the #s I have, and *definitely* not *nearly* enough to describe Supabase as "generally pretty far behind Deno Deploy and CF Workers"

Maybe it is just my setup being relatively ideal: I'm in Boston, and I can imagine ex. Tokyo might not have the same advantages.

That all being said, if you're like me and stumble across this thread, its *very* unlikely that there's an obvious, meaningful, win you'll get from trying Cloudflare Workers, and they're much more complex to setup IMHO, though not the worst thing in the world. I got ~95% of my latency improvement from silly, obvious, things, like running the request to check if a user is subscribed in parallel with the request to retrieve the user's balance.


If you're working at Supabase, I just wanna say thank you for all the work you put in. I'm honestly a bit stunned that there wasn't a meaningful difference from Cloudflare, and I can't tell you enough how big of an enabler and force multiplier your work was for me. I had 0 idea what I was doing, even know obvious things like "Whats CORS?" and "I *need* an edge function unless I'm planning on letting people using the web app see my API keys"

**Reply by shawnesquivel** [reply]
> TL;DR: Things are looking great, I don't see any speed difference between Cloudflare and Supabase, if there is, its < 100 ms AFAICT. If you stumble across this thread and are trying to figure out if you need to take some action, I strongly believe you don't.
> 
> Longer version:
> 
> My experience has been 99% mobile dev for the last 12 years, and over the last year of building out a startup, I've _heavily_ relied on Supabase to avoid having to learn server stuff.
> 
> I'm getting close to release, and about 6 weeks ago, started looking into various GitHub and Reddit threads about edge functions being slow.
> 
> I instrumented all my edge functions with Sentry traces, giving me precise breakdowns of slow spans of code.
> 
> I don't have a ton of users or huge set of data, its just me, the CI server, and a handful of people, definitely < 10. That being said, it's very clear to me at this point that Cloudflare has marginal speed gains, at best, in the context of proxying API calls to 3rd party services. Think, calling OpenAI, or a search result API. I'm talking about the search results API in the following, because its very steady traditionally, not much variance.
> 
> The very latest changes mentioned kept P25 response times ~the same, P50 got probably a 150 ms boost, P75, 200, broadening out to about 600 ms improvement at P99.
> 
> I just spent a couple hours wrapping up my first Cloudflare edge function that also does search. My instinct, with some charity to the above posts, is that if I deployed this I might see as much as a 100 ms improvement. That being said, there's no meaningful difference in speed in the #s I have, and _definitely_ not _nearly_ enough to describe Supabase as "generally pretty far behind Deno Deploy and CF Workers"
> 
> Maybe it is just my setup being relatively ideal: I'm in Boston, and I can imagine ex. Tokyo might not have the same advantages.
> 
> That all being said, if you're like me and stumble across this thread, its _very_ unlikely that there's an obvious, meaningful, win you'll get from trying Cloudflare Workers, and they're much more complex to setup IMHO, though not the worst thing in the world. I got ~95% of my latency improvement from silly, obvious, things, like running the request to check if a user is subscribed in parallel with the request to retrieve the user's balance.
> 
> If you're working at Supabase, I just wanna say thank you for all the work you put in. I'm honestly a bit stunned that there wasn't a meaningful difference from Cloudflare, and I can't tell you enough how big of an enabler and force multiplier your work was for me. I had 0 idea what I was doing, even know obvious things like "Whats CORS?" and "I _need_ an edge function unless I'm planning on letting people using the web app see my API keys"

Appreciate this thoughtful reply as someone who recently got a `InvalidWorkerCreation` and had the immediate thought that this wasn't production ready.

**Reply by laktek** [reply]
@shawnesquivel Did you get the `InvalidWorkerCreation` error on production or locally? I reached out to you on Twitter too. Would appreciate it if you open a ticket with your project details so I can look into the logs. It's uncommon to run into this error on a hosted project.
